### PR TITLE
Fix in Flutter Tools: Added kebab/param-cased as fileNames expected by the tool for multidimesional flavors in aab's build process 

### LINKED ISSUE #65264: Gradle build failed to produce an .aab file when multiple flavorDimensions specified
I was having trouble building a flutter app with `2 flavor dimensions` and `2 flavors` for each dimension (total of 4 variants, 
excluding the debug and release modes).

here's a minimal flutter app repo: https://github.com/brightknight08/flutter_build_appbundle_error to reproduce

when I do `flutter build appbundle --release --flavor signedprod`, it would display 
```clli

`flutter doctor -v`
```cli
PS E:\lab\flutter_build_appbundle_error> flutter doctor -v
[✓] Flutter (Channel stable, 1.20.2, on Microsoft Windows [Version 10.0.19041.450], locale en-US)
    • Flutter version 1.20.2 at C:\flutter
    • Framework revision bbfbf1770c (3 weeks ago), 2020-08-13 08:33:09 -0700
    • Engine revision 9d5b21729f
    • Dart version 2.9.1


[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at E:\asdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = E:\asdk
    • Java binary at: E:\as\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[✓] Android Studio (version 4.0)
    • Android Studio at E:\as
    • Flutter plugin version 48.1.2
    • Dart plugin version 193.7547
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[✓] VS Code (version 1.48.2)
    • VS Code at C:\Users\tim_m\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.14.0

[✓] Connected device (1 available)
    • P[redacted]...droid-arm64 • Android 10 (API 29)

• No issues found!
PS E:\lab\flutter_build_appbundle_error>
```

### PR DESCRIPTION
**Pull request by peupompeu**
This PR adds the kebab/param-cased version of the .aab bundle file name to the possible options so the gradle tool can look for it in the bundle directory and prevent the flutter build command failing because it could not find the .aab file and the end of the build.

List which issues are fixed by this PR. You must list at least one issue.
Fixes https://github.com/flutter/flutter/issues/65264

Based on @PritamSangani solution

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -6,6 +6,10 @@
     // reformatting of code.
     "editor.formatOnSave": false,
```
Comment (christopherfujino) [inline .vscode/settings.json:7]: Do we need these lines then?

Comment (christopherfujino) [reply .vscode/settings.json:7]: nvm, we probably don't want to be autoformatting anything in this repo

**Comment by christopherfujino** [comment]
@gnprice do you know if this is a reasonable solution?

File: .vscode/settings.json (hunk 1 above)
Comment (pepompeu) [reply .vscode/settings.json:7]: Vs-code doesn't respect the default rules of current .settings for the workspace scope, if you have especified some different behavior for ["dart"] in your global setup scope.

Added this lines to ensure respect for the original rules, avoiding auto formatting.

**Comment by gnprice** [comment]
This certainly looks plausible as the sort of thing I'd expect AGP to do when picking this filename.

The last time I looked for documentation of this sort of thing, there really wasn't any at all. So unless that's changed, the only way to try to pin down AGP's behavior here is for someone to try things out — make a small example Android project, muck with the flavors in the build, and see what happens.

That or study the AGP implementation, I guess, but the empirical approach may be easier.

One case I'd be curious about is what happens when an individual flavor is more than one word long — e.g. there's a `fooBar` flavor combined with a `bazQuux` flavor, on two different dimensions. This code predicts, IIUC, that that would come out as a file like `…/bundle/fooBarBazQuuxRelease/app-foo-bar-baz-quux-release.aab`. Is that what AGP in fact picks?

Or for more of an exotic edge case: what happens when individual flavors have names that start with a capital letter, and/or contain an underscore? It looks like there's a test case for that just above the one this PR adds, with a single `Foo_Bar` flavor. I think that's not a conventional way to name these, though, so if correctly supporting those would be more work, then they don't matter as much as names like `fooBar`.

**Comment by pepompeu** [comment]
> This certainly looks plausible as the sort of thing I'd expect AGP to do when picking this filename.
> 
> The last time I looked for documentation of this sort of thing, there really wasn't any at all. So unless that's changed, the only way to try to pin down AGP's behavior here is for someone to try things out — make a small example Android project, muck with the flavors in the build, and see what happens.
> 
> That or study the AGP implementation, I guess, but the empirical approach may be easier.
> 
> One case I'd be curious about is what happens when an individual flavor is more than one word long — e.g. there's a `fooBar` flavor combined with a `bazQuux` flavor, on two different dimensions. This code predicts, IIUC, that that would come out as a file like `…/bundle/fooBarBazQuuxRelease/app-foo-bar-baz-quux-release.aab`. Is that what AGP in fact picks?
> 
> Or for more of an exotic edge case: what happens when individual flavors have names that start with a capital letter, and/or contain an underscore? It looks like there's a test case for that just above the one this PR adds, with a single `Foo_Bar` flavor. I think that's not a conventional way to name these, though, so if correctly supporting those would be more work, then they don't matter as much as names like `fooBar`.

I'm interested in learning more about how the AGP behaves in different scenarios when it comes to naming files for projects with flavors. I'm willing to try out different use cases and experiment with various configurations. I'm going to try out the specific use case you mentioned with the fooBar and bazQuux flavors on two or more different dimensions to see what output file names are generated. I'll also look into the edge case you mentioned with flavors that have names starting with capital letters or containing underscores. Once I have some results to share, I'll provide a follow-up with my findings. Thanks again for your help!

**Comment by pepompeu** [comment]
Tested my proposal for this use cases:
```
    productFlavors {
        one {
            dimension "one"
        }
        two {
            dimension "two"
        }
        three {
            dimension "three"
        }
    }
```
```
    productFlavors {
        dimensionOne {
            dimension "one"
        }
        dimensionTwo {
            dimension "two"
        }
        dimensionThree {
            dimension "three"
        }
    }
```
```
    productFlavors {
        dimension_one {
            dimension "one"
        }
        dimension_two {
            dimension "two"
        }
        dimension_three {
            dimension "three"
        }
    }
```

In the current master the AGP exports the folders and files as follows:

camelCase file name: 
app-one-two-three-release.aab
app-dimensionOne-dimensionTwo-dimensionThree-release.aab (fooBar example)

snake_case file name: 
app-dimension_one-dimension_two-dimension_three-release.aab

In the latest master we get this error for the three file names:

> Gradle build failed to produce an .aab file. It's likely
> that this file was generated under
> /home/pedro/Desktop/flutter_lab/build, but the tool
> couldn't find it.

With my proposal we got a pass for the simple multidimensional flavors and snake_case flavor.
But still receives the error for camelCased flavor names ( fooBar example ).

As the flavor received from the BuildInfo class is only a string, we cant separate the dimensions of the flavor in fooBar situations 
ex: --flavor dimensionOneDimensionTwoDimensionThree, cant distinct the dimensions separators.

I would like to propose to add this restriction for flavor naming in flutter to only  lowercase and snakeCase until we develop a better way to interpret the --flavor tag in the build command

@gnprice what do you think we should do next?

**Comment by andrewkolos** [comment]
Thanks for the PR! Forgive me if I am misunderstanding, but I don't see how this fixes #65264. The issue (and the title of this PR) are about computing the correct .aab filename when building appbundles with multidimensional flavors, but the description and code change of this PR seems to be about an issue that occurs when the project has flavors with camelCased names (nothing to do with multidimensional flavors). This separate issue was mentioned as a comment on #65264, but I personally could not reproduce it on master on neither my MacBook nor my Ubuntu VM.

Regardless, we need a way to separate out each of the flavors per dimension (as @pepompeu pointed out). Here's one possibility I can think of: modify [packages/flutter_tools/gradle/src/main/groovy/flutter.groovy](https://github.com/flutter/flutter/blob/55c988fb453eddd97cb2e4d6df9ec28d72e2e899/packages/flutter_tools/gradle/src/main/groovy/flutter.groovy) to add a new task to print [productFlavors](https://developer.android.com/reference/tools/gradle-api/7.1/com/android/build/api/variant/ComponentIdentity#productFlavors:kotlin.collections.List). Then, modify [packages/flutter_tools/lib/src/android/gradle.dart](https://github.com/flutter/flutter/blob/55c988fb453eddd97cb2e4d6df9ec28d72e2e899/packages/flutter_tools/lib/src/android/gradle.dart) to invoke the task added in step 1 and read the output to get the separate flavors.

For inspiration, you can look at some code that does something similar for getting the build variants:
* in [flutter.goovy](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/gradle/src/main/groovy/flutter.groovy#L722)
* in [gradle.dart](https://github.com/flutter/flutter/blob/55c988fb453eddd97cb2e4d6df9ec28d72e2e899/packages/flutter_tools/lib/src/android/gradle.dart#L740)

**Comment by pepompeu** [comment]
Sorry for my confusion @andrewkolos, it's my first PR ever! Trying my best to help.

From my testing, the aab is built correctly and is placed in the output folder with the correct naming, in the correct place.

From my understanding, the problem is that the tool emits the warning reported in the issue if it can't find the outputed file.

The initial idea of the pr was to extend the files searched for the tool inside the folder to include this kind of outputs:

app-one-two-three-release.aab
app-dimensionOne-dimensionTwo-dimensionThree-release.aab (fooBar example)
app-dimension_one-dimension_two-dimension_three-release.aab

I'm missing something here?

**Comment by andrewkolos** [comment]
> From my testing, the aab is built correctly and is placed in the output folder with the correct naming, in the correct place.
> 
> From my understanding, the problem is that the tool emits the warning reported in the issue if it can't find the outputed file.

That's what I understand too. I think our difference in understanding is the problem that #65264 is describing. Focusing on only the issue description (first comment), I came up with this way to reproduce this issue:

1. Create a new flutter project.
2. In the `android` object android/app/build.gradle, add the following flavors:

```groovy
flavorDimensions "signed", "env"

productFlavors {
    signed {
        dimension "signed"
    }

    unsigned {
        dimension "signed"
    }

    prodThing { // We include some camel-casing to make sure this is not an issue.
        dimension "env"
    }

    internalThing {
        dimension "env"
    }
}
``` 
3. Run `flutter build appbundle --flavor signedprodThing`
4. You should see the error about the tool not being able to find the .aab.

Adding a `print` statement to the tool code, I see this is where the tool looks:
```
build/app/outputs/bundle/release/app.aab
build/app/outputs/bundle/release/app-release.aab
build/app/outputs/bundle/signedprodthingRelease/app.aab
build/app/outputs/bundle/signedprodthingRelease/app-signedprodthing-release.aab
build/app/outputs/bundle/signedprodThingRelease/app-signedprodThing-release.aab
```

However, the file that actually exists is
```
build/app/outputs/bundle/signedProdThingRelease/app-signed-prodThing-release.aab
\\                                                        ^ Notice the hyphen used to separate the flavors by dimension.
```

This PR would add the following path to search:
```
build/app/outputs/bundle/signedprodThingRelease/app-signedprod-thing-release.aab
```

which is still different.

**Comment by gnprice** [comment]
It seems like a sticky point here is that in a project like @andrewkolos's helpful example above,
* AGP gives that example "build variant" a name like `signedProdThingRelease`, with the two flavors (from the two flavor dimensions) and the build type (i.e., debug vs. release) all joined together in camel case. Docs here: https://developer.android.com/build/build-variants#flavor-dimensions
* If we have `--flavor` take a string like `signedProdThing`, then we can't tell whether that refers to (a) one flavor `signedProdThing`, (b) two flavors `signedProd` and `thing`, or (c) two flavors `signed` and `prodThing`. (In the example, the right answer is (c).)
* But we need to know the difference, because AGP also joins together the names in a different way, with hyphens. And we need both versions `signedProdThingRelease` and `signed-prodThing-release` as part of the output filename

One solution to that would be, as mentioned, to add a Gradle task that tells us what the flavors and flavor dimensions are, so that we can unpack a string like `signedProdThing`. That has a downside, though: running even a trivial Gradle task is expensive. For example in examples/hello_world/, if everything's already warm, running `cd android && ./gradlew help` takes 2-3 seconds on my Mac.

(This might be improved in the future by fixing #119196; not sure how much.)

Here's another possible solution: we could say that when you have multiple flavor dimensions, you need to express the different flavors to `--flavor` in some unambiguous form. E.g. we could say the flavors should be comma-separated:
```
flutter build appbundle --flavor signed,prodThing
```
Then there'd be no question that there are two flavors, namely `signed` and `prodThing`. From that we could compute both of the different ways of joining those names — `signedProdThing` and `signed-prodThing` — in order to successfully compute the filename `build/app/outputs/bundle/signedProdThingRelease/app-signed-prodThing-release.aab`.

Conveniently that would also have zero impact on everyone who's using `--flavor` for a single flavor dimension today: they'd keep saying `--flavor prodThing` or whatever, and it would unambiguously mean that there's one flavor `prodThing`.

**Comment by gnprice** [comment]
(I should also acknowledge, since I was somewhat negative above at https://github.com/flutter/flutter/pull/124797#issuecomment-1514164543 about Android's level of documentation of this sort of thing in AGP, that the docs have improved! That page https://developer.android.com/build/build-variants#flavor-dimensions is quite helpful, and even spells out most of what the filenames look like. I don't think that was there a year or two ago; and in fact the Internet Archive [first spotted that page](https://web.archive.org/web/20230418123907/https://developer.android.com/build/build-variants) just last week… indeed just a couple of hours after my comment.

Also although that page gives a very helpful example, it doesn't quite pin down the behavior to the degree one wants when writing code like this to interoperate with it. So the empirical investigations above are still essential.)

**Comment by christopherfujino** [comment]
@pepompeu thanks for your willingness to contribute! However, it seems like the approach in this PR is not the solution we want for https://github.com/flutter/flutter/issues/65264. I think further discussion on the correct approach should happen on that issue.

**Pull request closed without merging**